### PR TITLE
Add Tests for Reusable Filter Component

### PR DESCRIPTION
**Pull request by AnujChhikara**
Date: 29 March 2025

Developer Name: @AnujChhikara 

---

## Issue Ticket Number

- #987 
## Description

This PR adds test coverage for the reusable Filter component.

- A dedicated test file has been created to validate the core functionality of the Filter component, including UI rendering and button event listeners.

- Since the Filter component is used across four different pages, we've also added tests in each respective page's test file to ensure that the component renders correctly and displays the appropriate content in each context.

Feature PRs

- #988 
- #989

### Documentation Updated?

- [ ] Yes
- [x] No

### Under Feature Flag

- [x] Yes
- [ ] No

### Database Changes

- [ ] Yes
- [x] No

### Breaking Changes

- [ ] Yes
- [x] No

### Development Tested?

- [x] Yes
- [ ] No

## Screenshots
<details>
<summary>Screenshot 1</summary>

- Application Page

https://github.com/user-attachments/assets/59bfa0a0-f707-4b47-9fac-67fe7d871a06

- Extension Requests Page

https://github.com/user-attachments/assets/641fa769-f72a-49b4-b3f5-39e0d5ccb591

- Requests Page

https://github.com/user-attachments/assets/b4983e32-77ef-485d-a1aa-948aa4ebc9f5

- Task-Requests Page

https://github.com/user-attachments/assets/402058f7-537a-4c5c-8b95-27174c05f494

</details>

## Test Coverage
<details>
<summary>Screenshot 1</summary>

image
image

image
image
image

</details>

### CODE REVIEW

File: __tests__/task-requests/task-request.test.js [hunk 1]
```diff
@@ -42,7 +42,9 @@ describe('Task Requests', () => {
         });
       } else if (
         url ===
-        `${STAGING_API_URL}/taskRequests?size=20&q=status%3Aapproved++sort%3Acreated-asc`
+          `${STAGING_API_URL}/taskRequests?size=20&q=status%3Aapproved++sort%3Acreated-asc` ||
+        url ===
+          `${STAGING_API_URL}/taskRequests?size=20&q=status%3Aapproved+sort%3Acreated-asc`
```
Comment (AnujChhikara) [inline __tests__/task-requests/task-request.test.js:47]: The reason for adding this URL here is because it's called when approving a task request. Even though both URLs look almost identical, there's a small difference — one uses a single + before sort, and the other uses two +. So we included both to handle that minor variation and avoid any issues.

File: __tests__/extension-requests/extension-requests.test.js [hunk 2]
```diff
@@ -1209,4 +1209,26 @@ describe('Tests the Extension Requests Screen', () => {
       'Extension request successfully updated.',
     );
   });
+
+  it('should show only "APPROVED" and "DENIED" extension requests after selecting those filters when dev=true', async () => {
```
Comment (Suvidh-kaushik) [inline __tests__/extension-requests/extension-requests.test.js:1213]: if this test checks for only APPROVED and DENIED why are we checking all extensionRequests.length at the end won't it also contrain PENDING requests

**Review by Suvidh-kaushik — CHANGES_REQUESTED** [review]

File: __tests__/extension-requests/extension-requests.test.js (hunk 2 above)
Comment (AnujChhikara) [reply __tests__/extension-requests/extension-requests.test.js:1213]: So when that API is called to get the approved and pending requests, it responds with the mock extensionRequests data. That's why checking extensionRequests.length

Comment (Suvidh-kaushik) [reply __tests__/extension-requests/extension-requests.test.js:1213]: but the test is to show only approved and denied requests and won't the extensionRequests.length you are checking have all types of requests?

Comment (Suvidh-kaushik) [reply __tests__/extension-requests/extension-requests.test.js:1213]: but the test is to show only approved and denied requests and won't the extensionRequests.length you are checking have all types of requests?

File: __tests__/applications/applications.test.js [hunk 3]
```diff
@@ -361,4 +361,20 @@ describe('Applications page', () => {
...
+    const applyFilterButton = '[data-testid="apply-filter-component-button"]';
+    await page.waitForSelector(applyFilterButton, { visible: true });
+
+    await page.click(`input[type="checkbox"][id="ACCEPTED"]`);
+
+    await page.click(applyFilterButton);
+
+    await page.waitForNetworkIdle();
+    const applicationCardElements = await page.$$('.application-card');
+    expect(applicationCardElements.length).toBe(acceptedApplications.length);
```
[the remark this answers is not in the document]
Comment (AnujChhikara) [reply __tests__/applications/applications.test.js:378]: yes

[the remark this answers is not in the document]
Comment (ZendeAditya) [reply __tests__/applications/applications.test.js:378]: Okay

File: __tests__/components/filter.test.js [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+    const isModalHidden = await page.$eval(filterModalSelector, (el) =>
+      el.classList.contains('hidden'),
+    );
+    expect(isModalHidden).toBe(false);
+    const applyButton = await page.$(applyFilterButtonSelector);
+    expect(applyButton).toBeTruthy();
+    const clearButton = await page.$(clearFilterButtonSelector);
+    expect(clearButton).toBeTruthy();
+    const checkbox = await page.$(`input[type="checkbox"][id="ACCEPTED"]`);
+    expect(checkbox).toBeTruthy();
```
Comment (ZendeAditya) [inline __tests__/components/filter.test.js:92]: Shouldn't we display all three options in the modal instead of just `ACCEPTED` when the user opens it?

Comment (ZendeAditya) [reply __tests__/components/filter.test.js:92]: Or is this because we are on the `/application` page? As you write in the code, do we only show `ACCEPTED/APPROVED` on the `application` page?

File: __tests__/components/filter.test.js [hunk 5]
```diff
@@ -0,0 +1,266 @@
...
+  it('should apply filter, update URL, and display the corresponding filter tag', async () => {
+    const statusToApply = 'ACCEPTED';
+
+    await page.click(filterButtonSelector);
+    await page.waitForSelector(`${filterModalSelector}:not(.hidden)`);
+
+    await page.click(`input[type="checkbox"][id="${statusToApply}"]`);
+
+    await page.click(applyFilterButtonSelector);
+    await page.waitForSelector(filterModalSelector + '.hidden');
```
Comment (ZendeAditya) [inline __tests__/components/filter.test.js:129]: Can't we write simply like `await page.waitForSelector(`${filterModalSelector}.hidden`);
`?? Instead of the concatenation of the string, which is a little bit difficult to understand a little bit?

File: __tests__/components/filter.test.js [hunk 6]
```diff
@@ -0,0 +1,266 @@
...
+    await page.waitForNetworkIdle();
+
+    expect(page.url()).toContain(`status=${statusToApply.toLowerCase()}`);
+
+    const filterTagSelector = `${activeFilterTagsSelector} .filter__component__tag[data-status="${statusToApply}"]`;
```
Comment (ZendeAditya) [inline __tests__/components/filter.test.js:134]: Please do like this so it looks consistence.

File: __tests__/components/filter.test.js [hunk 7]
```diff
@@ -0,0 +1,266 @@
...
+  });
+
+  it('should clear filter via Clear button', async () => {
+    const statusToApply = 'PENDING';
+
+    await page.click(filterButtonSelector);
+    await page.waitForSelector(`${filterModalSelector}:not(.hidden)`);
+    await page.click(`input[type="checkbox"][id="${statusToApply}"]`);
+    await page.click(applyFilterButtonSelector);
+    await page.waitForSelector(filterModalSelector + '.hidden');
```
Comment (ZendeAditya) [inline __tests__/components/filter.test.js:151]: if possible please update here also

File: __tests__/components/filter.test.js [hunk 8]
```diff
@@ -0,0 +1,266 @@
...
+    await page.waitForNetworkIdle();
+    expect(page.url()).toContain(`status=${statusToApply.toLowerCase()}`);
```
Comment (ZendeAditya) [inline __tests__/components/filter.test.js:153]: After clearing the filter(Approved), the URL should not contain status=? filter?

Comment (AnujChhikara) [reply __tests__/components/filter.test.js:153]: @ZendeAditya where are we clearing the filter ??

File: __tests__/components/filter.test.js (hunk 4 above)
Comment (AnujChhikara) [reply __tests__/components/filter.test.js]: fixed thank you for pointing it out

File: __tests__/components/filter.test.js (hunk 5 above)
Comment (AnujChhikara) [reply __tests__/components/filter.test.js]: done

File: __tests__/components/filter.test.js (hunk 7 above)
Comment (AnujChhikara) [reply __tests__/components/filter.test.js]: done

File: __tests__/requests/requests.test.js [hunk 9]
```diff
@@ -585,4 +600,24 @@ describe('Tests the request cards', () => {
...
+    const applyFilterButton = '[data-testid="apply-filter-component-button"]';
+    await page.waitForSelector(applyFilterButton, { visible: true });
+    await page.click(`input[type="checkbox"][id="APPROVED"]`);
+    await page.click(applyFilterButton);
+    await page.waitForNetworkIdle();
+
+    const requestCards = await page.$$('[data-testid="ooo-request-card"]');
+    expect(requestCards.length).toBe(approvedRequestsData.data.length);
+
+    const statusText = await requestCards[0].$eval(
```
Comment (ZendeAditya) [inline __tests__/requests/requests.test.js:617]: Hey Anuj, why the check for only the first card instead of all? May `ooo-request-card` other request card also?

File: __tests__/components/filter.test.js [hunk 10]
```diff
@@ -0,0 +1,275 @@
...
+    const isClearButtonEnable = await page.$eval(
+      clearFilterButtonSelector,
+      (el) => !el.disabled,
+    );
+    expect(isClearButtonEnable).toBe(true);
+  });
+
+  it('should apply only one filter on application page', async () => {
+    await page.click(filterButtonSelector);
+    await page.waitForSelector(`${filterModalSelector}:not(.hidden)`);
```
Comment (ZendeAditya) [inline __tests__/components/filter.test.js:252]: The line `await page.waitForSelector(`${filterModalSelector}:not(.hidden)`);` appears more than ten times in the test file. Could you create a separate utility function for this? For example:

```javascript
const waitForFilterModal = async (page: Page, selector: string) => {
  await page.waitForSelector(`${selector}:not(.hidden)`);
};
```

Then, you can call this function each time with:

```javascript
await waitForFilterModal(page, filterModalSelector);
```

Thank you!

File: __tests__/requests/requests.test.js (hunk 9 above)
Comment (AnujChhikara) [reply __tests__/requests/requests.test.js:617]: We could do that, but it would be redundant since it would require looping through each card and checking them individually, which feels unnecessary in this case

File: __tests__/components/filter.test.js [hunk 11]
```diff
@@ -0,0 +1,275 @@
...
+      `input[type="checkbox"][id="ACCEPTED"]`,
+      (el) => el.checked,
+    );
+    expect(isAcceptedChecked).toBe(true);
+
+    isPendingChecked = await page.$eval(
+      `input[type="checkbox"][id="PENDING"]`,
+      (el) => el.checked,
+    );
+    expect(isPendingChecked).toBe(false);
```
Comment (ZendeAditya) [inline __tests__/components/filter.test.js:273]: Why don't you add `Rejected`? If somebody selects `Reject`, then? add a block for check `Reject`
check

Comment (AnujChhikara) [reply __tests__/components/filter.test.js:273]: If we add a separate check for "Rejected", then we’d also need to add similar blocks for every other status. Since the logic remains the same for each scenario, adding multiple explicit checks would be unnecessary and repetitive

File: __tests__/requests/requests.test.js (hunk 9 above)
Comment (ZendeAditya) [reply __tests__/requests/requests.test.js:617]: But what if it shows other cards instead of `Approved`?

Comment (AnujChhikara) [reply __tests__/requests/requests.test.js:617]: In line 615 we are already checking if the card we are rendering are equal to the api response of approved requests

Comment (ZendeAditya) [reply __tests__/requests/requests.test.js:617]: Okay then there is no problem

File: __tests__/components/filter.test.js [hunk 12]
```diff
@@ -0,0 +1,275 @@
...
+
+    const acceptedStatus = 'ACCEPTED';
+    const tag = await page.$(
+      `${activeFilterTagsSelector} .filter__component__tag[data-status="${acceptedStatus}"]`,
+    );
+    expect(tag).toBeTruthy();
+
+    await page.click(filterButtonSelector);
+    await page.waitForSelector(`${filterModalSelector}:not(.hidden)`);
+    const isCheckboxChecked = await page.$eval(
```
Comment (pankajjs) [inline __tests__/components/filter.test.js:238]:
```suggestion
    const isAcceptedChecked = await page.$eval(
```

File: __tests__/components/filter.test.js [hunk 13]
```diff
@@ -0,0 +1,275 @@
...
+      `input[type="checkbox"][id="${acceptedStatus}"]`,
+      (el) => el.checked,
+    );
+    expect(isCheckboxChecked).toBe(true);
+    const isClearButtonEnable = await page.$eval(
```
Comment (pankajjs) [inline __tests__/components/filter.test.js:243]:
```suggestion
    const isClearButtonEnabled = await page.$eval(
```

**Comment by pankajjs** [comment]
Please update the test screenshots

File: __tests__/task-requests/task-request.test.js (hunk 1 above)
Comment (MayankBansal12) [reply __tests__/task-requests/task-request.test.js:47]: instead of using hardcoded urls, can we use `URLSearchParams` here?
it will help you avoid this situation

Comment (AnujChhikara) [reply __tests__/task-requests/task-request.test.js:47]: since the rest of the tests match full URLs as strings. Didn't want to overcomplicate for one edge case — can always refactor later if needed.

**Review by ZendeAditya — APPROVED** [review]

**Review by surajmaity1 — APPROVED** [review]
LGTM

**Review by iamitprakash — APPROVED** [review]
